### PR TITLE
Draft: params.clones

### LINKED ISSUE #30: clone & reset handling different than feathers-vuex
In `feathers-vuex` cloning a clone isn't possible. `item.clone().clone()` throws an error: `'You cannot clone a copy'`.

```ts
// see https://github.com/feathersjs-ecosystem/feathers-vuex/blob/8643bd7b0cd045fcfd1e407143c3221ae58b7486/src/service-module/make-base-model.ts#L305-L313
if (this.__isClone) {
  throw new Error('You cannot clone a copy')
}
```

In `feathers-pinia` cloning a clone is totally possible. You already have a test for it. See:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/tests/model-instance-clone-and-commit.test.ts#L39-L47

Don't know, if it's intended, though. In `feathers-pinia` there is `item.reset()` as in `feathers-vuex`, but it calls `store.resetCopy()`, which is not existent. See:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/base-model.ts#L131-L135

If `clone.clone()` is intended to work, that's a breaking change from `feathers-vuex` worth mentioning in the docs!

Also constructor adds `__isClone` differently compared to `feathers-vuex`. In `feathers-vuex` it only gets added when it should be a clone:
```ts
// see https://github.com/feathersjs-ecosystem/feathers-vuex/blob/8643bd7b0cd045fcfd1e407143c3221ae58b7486/src/service-module/make-base-model.ts#L151-L157
// Mark as a clone
if (options.clone) {
  Object.defineProperty(this, '__isClone', {
    value: true,
    enumerable: false
  })
}
```
In `feathers-pinia` it always gets added as `true` or `false`:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/base-model.ts#L24

You also make it explicitely defined in (see exclamation mark):
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/base-model.ts#L18

But on `clone.commit()` you remove it. You have a test for that:
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/tests/model-instance-clone-and-commit.test.ts#L27-L37

That's because `fast-copy` does not add the `enumerable: false` property of `__isClone`. Whats the intended way?
https://github.com/marshallswain/feathers-pinia/blob/19a13eec04605f140c1e5b4f46e13c132c0052d3/src/service-store/make-actions.ts#L343

Todos:
- [x] say if `clone.clone()` should work or not @marshallswain 
- [x] say if `clone.__isClone` should always return a boolean or if it is not existent @marshallswain 
- [x] get `clone.reset()` running again, add old `resetCopy` mutation as `resetClone` action @fratzinger 
- [x] add `resetClone` renaming to the 'What's Different from Feathers-Vuex' section
- [x] add a test for `clone.reset()` @fratzinger 
- [ ] If `clone.clone()` is intended to work, add it to the 'What's Different from Feathers-Vuex' section @fratzinger 
- [ ] ~If `clone.__isClone` should always return a boolean, add `Object.defineProperty` to `store.commit`; change the test of `commit()` accordingly @fratzinger~

### PR DESCRIPTION
**Pull request by fratzinger**
- [x] feat(findInStore): add params.clones
- [x] feat: findInStore with params.clones
- [x] feat: getFromStore with params.clones 
- [x] fix(findInStore): don't readd picked values
- [x] refactor(findInStore): move _.values to `items` / `temps` getters -> micro perf
- [x] refactor: simplify `addOrUpdate`
- [x] docs: add `params.clones`
- [x] tests
- [x] fix: add `idField` to `getAnyId`
- [x] feat: add micro-optimization getters (`itemsAndTemps`, `itemsAndClones`, `itemsTempsAndClones`, `operators`)
- [x] perf: `findInStore` micro-optimized
- [x] fix: `model.reset()` working again
- [x] add `resetClone` renaming to the 'What's Different from Feathers-Vuex' section
- [x] add a test for `clone.reset()` @fratzinger 
- [ ] add new getters to docs
- [ ] add `clone.clone()` to the 'What's Different from Feathers-Vuex' section

Because `feathers-pinia` uses `clones` over `copies` consistently, I decided to call it `params.clones`. It's a (small) breaking change compared to `feathers-vuex`.

Also I also added variable `pickedValues` because otherwise if `$select` is used, the items get added and added to the store. That's also problematic for clones, because the cloned $selected values would overwrite the `itemsById` values. I'm happy to discuss this.

closes #30

### CODE REVIEW

File: src/service-store/make-getters.ts [hunk 1]
```diff
@@ -78,8 +85,12 @@ export function makeGetters(options: ServiceOptions): ServiceGetters {
           values = values.slice(0, filters.$limit)
         }
 
+        // keep transformed items by $select separately
```
Comment (J3m5) [inline src/service-store/make-getters.ts:88]: I think we shouldn't can  addOrUpdate in the getter as it will cause side effects and that is generally discouraged.
Posva himself have also insisted on that point [there](https://github.com/posva/pinia/discussions/530#discussioncomment-877407
)
I'm currently doing a rather big refactoring and once the store getters, actions and state are all together in the same object, we  get back state types and all and typescrypt complain about that because actions are normally not available on the `this` type in getters.

https://github.com/posva/pinia/discussions/709#discussioncomment-1449092

There isn't any clear way of splitting stores by state, getters and actions yet:
https://github.com/posva/pinia/issues/802#issuecomment-971390377
https://github.com/posva/pinia/issues/829#issue-1062161077

Comment (fratzinger) [reply src/service-store/make-getters.ts:88]: It's just for SSR and was discussed before. It makes perfectly sense for SSR. https://feathers-pinia.pages.dev/guide/service-stores.html#automatic-instance-hydration
We need to find a way that it works with `$select`. The downside of my current solution is, that it doesn't return a Model instance which is not reflected by the types

Comment (J3m5) [reply src/service-store/make-getters.ts:88]: Then shouldn't we do it only when we are in an SSR environment ?
https://pinia.esm.dev/ssr/#state-hydration

Comment (marshallswain) [reply src/service-store/make-getters.ts:88]: I actually implemented it with full knowledge of the violation of this cardinal sin.  :)  Since the possible negative impact is very low, I consider the API to be acceptable. The possible impact is only one additional render for the entire list.  In order for it to be a problem, the user would need to loop through the results and call a model instance method immediately upon render.  It's a low likelihood that somebody will create an app that requires that.... BUT we require a solution for that scenario.

The location that you pointed out is actually a cleaner implementation, but potentially less performant for apps that pre-load a large number of records.  Converting a large number of records into instances can cause a noticeable delay in loading data for Vue 2.  In Vue 3 I imagine it's going to be barely noticeable.  The existing setup is basically the lazy hydration implementation, if a bit dirty.

So basically, I don't plan on removing the call to `addOrUpdate`, but we do require a solution for those who would prefer a cleaner setup.  A utility to hydrate store data into instances during pinia hydration (https://pinia.esm.dev/ssr/#state-hydration) would be perfect.

**Comment by marshallswain** [comment]
This is looking solid.  Could you add some tests to make the spec official?

**Comment by fratzinger** [comment]
Currently at it ;)

**Comment by fratzinger** [comment]
I think it's ready now. I also added `params.clones` support for `getFromStore`. This diverges from `findInStore` a little bit. 
In `findInStore` we have `temps: true` and `clones: true`. In `getFromStore` we don't have `temps: true`. Instead we can use `getFromStore(__tempId)`.
So there are two behaviors:
- `findInStore({ query: {}, temps: false, clones: true }` returns just items with `id` and and if existent, their cloned items
- For `getInStore` there no option to set `temps: false`

Maybe we can just leave it like that. Maybe we should add support for `temps: false` for `getFromStore`.

**Comment by fratzinger** [comment]
We could even do more micro optimizations, if we have a `itemsAndTemps`, `itemsWithClones`, `itemsAndTempsWithClones` store getters. In `findInStore` we could use:
```js
let values
if (!params.temps && !params.clones) {
  values = this.items
} else if (params.temps && !params.clones) {
  values = this.itemsAndTemps
} else if (!params.temps && params.clones) {
  values = this.itemsWithClones
} else {
  values = this.itemsAndTempsWithClones
}
```
The more a `findInStore` is used with temps or clones, the more the micro optimization pays off because the concatenation of `items`, `temps` and `clones` is cached.
What do you think?

**Comment by marshallswain** [comment]
I like it.  Let’s call them

itemsAndTemps
itemsAndClones
itemsTempsAndClones

On January 18, 2022, GitHub ***@***.***> wrote:
> We could even do more micro optimizations, if we have a itemsAndTemps,
> itemsWithClones, itemsAndTempsWithClones store getters. In findInStore
> we could use:
>
> let values if (!params.temps && !params.clones) { values = this.items
> } else if (params.temps && !params.clones) { values =
> this.itemsAndTemps } else if (!params.temps && params.clones) { values
> = this.itemsWithClones } else { values = this.itemsAndTempsWithClones
> }
> The more a findInStore is used with temps or clones, the more the
> micro optimization pays off because the concatenation of items, temps
> and clones is cached.
>  What do you think?
>

—
Reply to this email directly, view it on GitHub
<https://github.com/marshallswain/feathers-pinia/pull/24#issuecomment-
1015937388>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAA7OWJQTWHKY2DAZAGJ733UWX4EPANCNFSM5JZFGDTQ>.
You are receiving this because you commented.Message ID:
***@***.***>

**Comment by fratzinger** [comment]
Started it and went down the rabbit hole in #30. Need decisions first over there (see the todos at the bottom of the issue). When decisions are clear, I'll add the other todos to this PR.

**Comment by fratzinger** [comment]
Reason for #30 was a failing test. I don't know what to do. There is a quirk about `__isClone` being `false` and `undefined`. Please @marshallswain could you have a look?:

- test: https://github.com/fratzinger/feathers-pinia/blob/2d4d66ac04e86291535d38038ad878c13c0bd79d/tests/lists.test.ts#L104-L125
- test result: https://github.com/marshallswain/feathers-pinia/runs/4908931477?check_suite_focus=true

**Comment by marshallswain** [comment]
@fratzinger, I'm still open to adding this feature.  Since this PR is so far outdated, now, I'm going to go ahead and close it and we can re-branch from `v2`.

**Comment by marshallswain** [comment]
I went ahead and added the querying part of the feature through `findInStore` and `getFromStore`: https://github.com/marshallswain/feathers-pinia/releases/tag/v2.0.0-pre.11
